[CDAP-18694] Allow preview runner to fetch metadata from AppFab and use sidecar for fetching and caching artifacts

wyzhang opened the pull request:
What:
When artifact localizer is enabled, support preview runner to
fetch metadata and artifacts from AppFab instead of directly
from metadata DB (i.e. sql) and from distributed file system.

It actives this by making preview runner
* fetching metadata from AppFab via remote implementations of artifact repository, plugin finder, preference
* fetching artifact from AppFab via artifact localizer running in sidecar container.

Why:
To allow preview runner running with restricted permission (i.e.
no permission to access artifacts on distributed file system or
metadata in external database).

sagarkapare commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepository.java:
```diff
@@ -50,21 +55,34 @@
+    } else {
+      location = artifactDescriptor.getLocation();
+    }
+    return artifactClassLoaderFactory.createClassLoader(location, entityImpersonator);
```
minor nit: maybe better if localizeArtifact is responsible for returning location so lines 78-84 can be moved to localizeArtifact method and this can simply be

```java
public CloseableClassLoader createArtifactClassLoader(ArtifactDescriptor artifactDescriptor,
                                                        EntityImpersonator entityImpersonator) throws IOException {
   Location location = localizeArtifact(artifactId, artifactDescriptor.getLocation());
  return artifactClassLoaderFactory.createClassLoader(location, entityImpersonator);
}
```

wyzhang replied:
Valid point. 

Restructure the code differently and looks to be cleaner. 

In short, 
created 
class xxWithLocalization extends xxx {
}

where 
class xxx {
Location getLocation() {}
}

and 
class xxWithLocalization {
Location getLocation() { 
  // localize and return local location
}

sagarkapare commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java:
```diff
@@ -153,10 +168,17 @@ public void close() throws IOException {
     for (ArtifactDetail detail : details) {
-      Location artifactLocation = locationFactory.create(detail.getDescriptor().getLocationURI());
+      Location location;
+      if (artifactLocalizerClient.isPresent()) {
```
minor nit: similarly here, if we have localizeArtifact method as  localizeArtifact(ArtifactId artifactId, Location defaultLocation), we can avoid this duplicate checks here and on line 103.

wyzhang replied:
Same as above

sagarkapare reviewed (approved):
minor nits else LGTM 👍

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewRunnerModule.java:
```diff
@@ -121,17 +132,51 @@
+        .annotatedWith(Names.named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO))
+        .to(RemoteArtifactRepository.class)
+        .in(Scopes.SINGLETON);
+      expose(ArtifactRepository.class).annotatedWith(Names.named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO));
```
nit: Do we need to expose annotated ArtifactRepository.class separately? In line 142, ArtifactRepository.class is already being exposed.

wyzhang replied:
I think we do need.

wyzhang committed:
[CDAP-18694] Allow preview runner to fetch metadta from AppFab and use sidecar for fetching and caching aritfacts

What:
When artifact localizer is enabled, support preview runner to
fetch metadata and artifacts from AppFab instead of directly
from metadata DB (i.e. sql) and from distributed file system.

It actives this by making preview runner
* fetching metadata from AppFab via remote implementations of artifact repository, plugin finder, preference
* fetching artifact from AppFab via artifact localizer running in sidecar container.

Why:
To allow preview runner running with restricted permission (i.e.
no permission to access artifacts on distributed file system or
metadata in external database).
